security/src/main/scala/pl/touk/nussknacker/ui/security/accesslogic/AnonymousAccess.scala

```diff
@@ -0,0 +1,28 @@
+package pl.touk.nussknacker.ui.security.accesslogic
+
+import akka.http.scaladsl.server.Directives.{handleRejections, reject}
+import akka.http.scaladsl.server.{AuthorizationFailedRejection, Directive0, RejectionHandler}
+import pl.touk.nussknacker.ui.security.api.{AuthenticatedUser, AuthenticationResources}
+
+import scala.concurrent.Future
+
+class AnonymousAccess(authenticationResources: AuthenticationResources) {
+
+  val anonymousUserRole: Option[String] = authenticationResources.configuration.anonymousUserRole
+
+  def handleAuthorizationFailedRejection: Directive0 = handleRejections(
+    RejectionHandler
+      .newBuilder()
+      // If the authorization rejection was caused by anonymous access,
+      // we issue the Unauthorized status code with a challenge instead of the Forbidden
+      .handle { case AuthorizationFailedRejection => authenticationResources.authenticate() { _ => reject } }
+      .result()
+  )
+
+  def authenticateWithAnonymousAccess(): Future[Option[AuthenticatedUser]] = anonymousUserRole
```

Comment:
WDYT about sth like this?
```scala
  def authenticateWithAnonymousAccess(): Future[Option[AuthenticatedUser]] = Future {
    anonymousUserRole.map(role => AuthenticatedUser.createAnonymousUser(Set(role)))
  }
```

and let's add `implicit executionContext: ExecutionContext` to the constructor. When you take a look at the Future.successful method def you will find sth like this:

```scala
final def successful[T](result: T): Future[T]
```
For some reason, it doesn't use `: =>` (by-name param) so the `T` will be evaluated before creating the Future. It means that all exception that will be thrown during evaluation won't be placed in the Future context (as Future.failed). So, this means that `Future.successful` should be used only when you have already evaluated value (or at least some trivial code that won't throw for sure)

In our case, we call the `createAnonymousUser` method. From this place, we cannot assume the code won't throw any exception. Using `Future.apply` solves the problem.

Reply:
Oh, I was not aware of that. Thank you! I modified it so potential errors will be placed in the context.